File: backend/src/main/java/kr/co/programmers/collabond/api/recruit/domain/RecruitPost.java
package kr.co.programmers.collabond.api.recruit.domain;

import jakarta.persistence.*;
import kr.co.programmers.collabond.api.profile.domain.Profile;
import kr.co.programmers.collabond.shared.domain.UpdatedEntity;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.SQLDelete;
import org.hibernate.annotations.SQLRestriction;

import java.time.LocalDateTime;

@Entity
@Table(name = "recruit_posts")
@Getter
@SQLDelete(sql = "UPDATE recruit_posts SET deleted_at = NOW() WHERE id = ?")
@SQLRestriction("deleted_at IS NULL")
@NoArgsConstructor(access = lombok.AccessLevel.PROTECTED)
public class RecruitPost extends UpdatedEntity {

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "profile_id")
    private Profile profile;

    @Column(nullable = false)
    private String title;

    @Column(nullable = false, length = 1000)
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private RecruitPostStatus status;

    @Column(nullable = false)
    private LocalDateTime deadline;

    @Builder
    private RecruitPost(Profile profile, String title, String description,
                        RecruitPostStatus status, LocalDateTime deadline) {

        this.profile = profile;
        this.title = title;
        this.description = description;
        this.status = status;
        this.deadline = deadline;
    }

    public RecruitPost update(String title, String description,
                              RecruitPostStatus status, LocalDateTime deadline) {

        if (title != null) {
            this.title = title;
        }

        if (description != null) {
            this.description = description;
        }

        if (status != null) {
            this.status = status;
        }

        if (deadline != null) {
            this.deadline = deadline;
        }

        return this;
    }
}

File: backend/src/main/java/kr/co/programmers/collabond/api/recruit/domain/RecruitPost.kt
package kr.co.programmers.collabond.api.recruit.domain

import jakarta.persistence.*
import kr.co.programmers.collabond.api.profile.domain.Profile
import kr.co.programmers.collabond.shared.domain.UpdatedEntity
import org.hibernate.annotations.SQLDelete
import org.hibernate.annotations.SQLRestriction
import java.time.LocalDateTime

@Entity
@Table(name = "recruit_posts")
@SQLDelete(sql = "UPDATE recruit_posts SET deleted_at = NOW() WHERE id = ?")
@SQLRestriction("deleted_at IS NULL")
class RecruitPost(
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "profile_id")
    var profile: Profile? = null,

    @Column(nullable = false)
    var title: String,

    @Column(nullable = false, length = 1000)
    var description: String,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var status: RecruitPostStatus = RecruitPostStatus.RECRUITING,

    @Column(nullable = false)
    var deadline: LocalDateTime
) : UpdatedEntity() {

    protected constructor() : this(
        profile = null,
        title = "",
        description = "",
        deadline = LocalDateTime.now()
    )

    fun update(
        title: String? = null,
        description: String? = null,
        status: RecruitPostStatus? = null,
        deadline: LocalDateTime? = null
    ): RecruitPost = apply {
        title?.let { this.title = it }
        description?.let { this.description = it }
        status?.let { this.status = it }
        deadline?.let { this.deadline = it }
    }
}